File: analysis/running_wsls/wsls_scatter_posterior.py
"""
Two-panel figure (style matches wm_wsls.png):
  Left  — scatter P(stay|win) vs P(stay|lose) per subject (all 18)
           Blue  #0072B2 = understood=felt (n=13)
           Orange #E69F00 = understood≠felt (n=5)
  Right — β_mu posterior from hierarchical WSLS on the 13 included subjects
"""

import os
import re
import warnings
import numpy as np
import pandas as pd
import pymc as pm
import arviz as az
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
from scipy.stats import linregress, gaussian_kde

warnings.filterwarnings("ignore")
os.chdir(os.path.dirname(os.path.abspath(__file__)))

DATA_DIR = "../../data/ioc-all-fixed-pilot/task"
OUT_PNG  = "wsls_scatter_posterior.png"
DRAWS_NC = "wsls_13_draws.nc"

MISUNDERSTOOD  = {"67dae998d8f2cfb8a8e3bf03"}
FELT_DIFFERENT = {
    "677009b08130c3028f6a8a6d",
    "68598a1d4cebd213b2abb1d9",
    "69b7e04340b00585acbb91ac",
    "6a0092581cd317f1ff1765a2",
}
OMITTED = MISUNDERSTOOD | FELT_DIFFERENT

COL_INC   = "#0072B2"   # Okabe-Ito blue   — understood = felt
COL_OMT   = "#E69F00"   # Okabe-Ito orange — understood ≠ felt
COL_TREND = "#EE6677"   # Paul Tol rose
GREY_LINE = "#404040"
ZERO_COL  = "#666666"
GREY_MED  = "#888888"


# ── data ──────────────────────────────────────────────────────────────────────
def load_raw(data_dir):
    records = []
    for fname in sorted(os.listdir(data_dir)):
        if not fname.endswith(".csv"):
            continue
        m = re.match(r"ioc-all_([a-f0-9]{24})_SESSION", fname)
        if not m:
            continue
        pid = m.group(1)
        df  = pd.read_csv(os.path.join(data_dir, fname), low_memory=False)
        gb  = (
            df[
                (df["task"] == "gambling_choice") &
                (df["block_number"].astype(str) != "training") &
                (df["is_choice_valid"].astype(str).str.lower() == "true")
            ]
            .copy()
            .assign(block_number=lambda x: x["block_number"].astype(int))
            .sort_values(["block_number", "trial_number"])
            .reset_index(drop=True)
        )
        gb["reward"]          = pd.to_numeric(gb["reward"], errors="coerce")
        gb["trial_number"]    = pd.to_numeric(gb["trial_number"], errors="coerce")
        gb["is_choice_valid"] = gb["is_choice_valid"].astype(str).str.lower() == "true"
        records.append({"pid": pid, "df": gb})
    return records


def per_subject_stay_rates(records):
    rows = []
    for rec in records:
        gb = rec["df"].copy()
        gb["choice_prev"] = gb.groupby("block_number")["choice_key"].shift(1)
        gb["reward_prev"] = gb.groupby("block_number")["reward"].shift(1)
        gb["valid_prev"]  = gb.groupby("block_number")["is_choice_valid"].shift(1)
        sub = gb[
            (gb["trial_number"] > 1) &
            gb["is_choice_valid"] &
            gb["valid_prev"].fillna(False) &
            gb["choice_prev"].notna() &
            gb["reward_prev"].notna()
        ].copy()
        sub["stay"]        = (sub["choice_key"] == sub["choice_prev"]).astype(int)
        sub["reward_prev"] = sub["reward_prev"].astype(int)
        p_win  = float(sub.loc[sub["reward_prev"] == 1, "stay"].mean()) \
                 if (sub["reward_prev"] == 1).sum() > 0 else np.nan
        p_lose = float(sub.loc[sub["reward_prev"] == 0, "stay"].mean()) \
                 if (sub["reward_prev"] == 0).sum() > 0 else np.nan
        rows.append({"pid": rec["pid"], "p_win": p_win, "p_lose": p_lose,
                     "omitted": rec["pid"] in OMITTED})
    return pd.DataFrame(rows).dropna(subset=["p_win", "p_lose"])


def make_wsls_df(records):
    frames = []
    for rec in records:
        gb = rec["df"].copy()
        gb["participant"] = rec["pid"]
        frames.append(gb)
    raw = pd.concat(frames, ignore_index=True)
    raw = raw.sort_values(["participant", "block_number", "trial_number"])
    raw["choice_prev"]   = raw.groupby(["participant","block_number"])["choice_key"].shift(1)
    raw["reward_prev"]   = raw.groupby(["participant","block_number"])["reward"].shift(1)
    raw["is_valid_prev"] = raw.groupby(["participant","block_number"])["is_choice_valid"].shift(1)
    raw = raw[
        (raw["trial_number"] > 1) &
        raw["is_choice_valid"] &
        raw["is_valid_prev"].fillna(False) &
        raw["choice_prev"].notna() &
        raw["reward_prev"].notna()
    ].copy()
    raw["stay"]        = (raw["choice_key"] == raw["choice_prev"]).astype(int)
    raw["reward_prev"] = raw["reward_prev"].astype(int)
    pids    = sorted(raw["participant"].unique())
    pid_map = {p: i for i, p in enumerate(pids)}
    raw["subj_idx"] = raw["participant"].map(pid_map)
    return raw, pids


def fit_wsls(df, n_subjects):
    print(f"Fitting WSLS (N={n_subjects}, {len(df)} trials) …")
    with pm.Model():
        alpha_mu    = pm.Normal("alpha_mu",    0, 2)
        beta_mu     = pm.Normal("beta_mu",     0, 2)
        alpha_sigma = pm.HalfNormal("alpha_sigma", 1)
        beta_sigma  = pm.HalfNormal("beta_sigma",  1)
        alpha_z = pm.Normal("alpha_z", 0, 1, shape=n_subjects)
        beta_z  = pm.Normal("beta_z",  0, 1, shape=n_subjects)
        pm.Deterministic("alpha", alpha_mu + alpha_sigma * alpha_z)
        pm.Deterministic("beta",  beta_mu  + beta_sigma  * beta_z)
        log_odds = (
            (alpha_mu + alpha_sigma * alpha_z)[df["subj_idx"].values]
            + (beta_mu + beta_sigma * beta_z)[df["subj_idx"].values]
            * df["reward_prev"].values
        )
        pm.Bernoulli("stay", logit_p=log_odds, observed=df["stay"].values)
        idata = pm.sample(draws=1000, tune=1000, chains=4,
                          target_accept=0.9, progressbar=True, random_seed=42)
    return idata


# ── main ──────────────────────────────────────────────────────────────────────
def main():
    all_records = load_raw(DATA_DIR)

    # scatter data
    sdf    = per_subject_stay_rates(all_records)
    p_win  = sdf["p_win"].values
    p_lose = sdf["p_lose"].values
    colors = [COL_OMT if row["omitted"] else COL_INC
              for _, row in sdf.iterrows()]
    assert len(p_win) == len(p_lose)

    # WSLS posterior on 13 included subjects
    incl = [r for r in all_records if r["pid"] not in OMITTED]
    wsls_df, pids = make_wsls_df(incl)

    if os.path.exists(DRAWS_NC):
        print("Loading saved WSLS draws …")
        idata = az.from_netcdf(DRAWS_NC)
    else:
        idata = fit_wsls(wsls_df, len(pids))
        az.to_netcdf(idata, DRAWS_NC)

    beta_mu_draws = idata.posterior["beta_mu"].values.flatten()

    # ── figure ────────────────────────────────────────────────────────────────
    fig, (ax_sc, ax_post) = plt.subplots(1, 2, figsize=(10, 4.5))

    # ── left: scatter ─────────────────────────────────────────────────────────
    # axes switched: x = P(stay|lose), y = P(stay|win)
    shared_lo = 0.0
    shared_hi = max(p_win.max(), p_lose.max()) * 1.08

    # diagonal reference line (y = x)
    ax_sc.plot([shared_lo, shared_hi], [shared_lo, shared_hi],
               linestyle="--", color=GREY_LINE, linewidth=0.9, alpha=0.6, zorder=1)

    # OLS trend line
    slope, intercept, *_ = linregress(p_lose, p_win)
    x_fit = np.array([shared_lo, shared_hi])
    ax_sc.plot(x_fit, intercept + slope * x_fit,
               color=COL_TREND, linewidth=1.2, zorder=2)

    for i in range(len(sdf)):
        ax_sc.scatter(p_lose[i], p_win[i],
                      color=colors[i], s=50, alpha=0.85, zorder=3)

    ax_sc.set_xlim(shared_lo, shared_hi)
    ax_sc.set_ylim(shared_lo, shared_hi)

    legend_els = [
        Line2D([0],[0], marker="o", color="w", markerfacecolor=COL_INC,
               markersize=7, label="understood = felt  (n=13)"),
        Line2D([0],[0], marker="o", color="w", markerfacecolor=COL_OMT,
               markersize=7, label="understood ≠ felt  (n=5)"),
    ]
    ax_sc.legend(handles=legend_els, fontsize=7.5, loc="upper left",
                 framealpha=0.8, edgecolor="none")

    ax_sc.set_xlabel("P(stay | loss)", fontsize=10)
    ax_sc.set_ylabel("P(stay | win)", fontsize=10)
    ax_sc.tick_params(labelsize=8)
    ax_sc.spines[["top", "right"]].set_visible(False)

    # ── right: posterior ──────────────────────────────────────────────────────
    med  = float(np.median(beta_mu_draws))
    pd_  = max(np.mean(beta_mu_draws > 0), np.mean(beta_mu_draws < 0)) * 100
    ci90 = np.quantile(beta_mu_draws, [0.05, 0.95])

    kde  = gaussian_kde(beta_mu_draws, bw_method="scott")
    xs   = np.linspace(beta_mu_draws.min() * 1.15, beta_mu_draws.max() * 1.15, 512)
    ys   = kde(xs); ys = ys / ys.max()

    ax_post.fill_between(xs, ys, color="#c8c8c8", linewidth=0)
    ax_post.axvline(0,   color=ZERO_COL, linestyle="--", linewidth=0.9, alpha=0.8)
    ax_post.axvline(med, color=GREY_MED, linestyle="--", linewidth=0.5, alpha=0.7)

    CI_Y = -0.08
    ax_post.plot([ci90[0], ci90[1]], [CI_Y, CI_Y],
                 color=GREY_LINE, linewidth=1.0, solid_capstyle="round")
    ax_post.plot(med, CI_Y, "o", color=GREY_LINE, markersize=5, zorder=5)

    ax_post.text(med, 1.22,
                 f"median = {med:.2f}\npd = {pd_:.1f}%",
                 ha="center", va="top", fontsize=8, color="#555555")

    max_abs = max(abs(beta_mu_draws.min()), abs(beta_mu_draws.max())) * 1.1
    ax_post.set_xlim(-max_abs, max_abs)
    ax_post.set_ylim(CI_Y - 0.15, 1.5)

    ax_post.set_xlabel(r"$\beta_{\mu}$  (reward effect on log-odds of staying)",
                       fontsize=10)
    ax_post.set_title("WSLS reward effect\nFelt = instructed  (N = 13, 5 omitted)",
                      fontsize=10, pad=6)
    ax_post.spines[["top", "right", "left"]].set_visible(False)
    ax_post.yaxis.set_visible(False)
    ax_post.spines["bottom"].set_color("#888888")
    ax_post.tick_params(axis="x", labelsize=8)

    plt.tight_layout()
    plt.savefig(OUT_PNG, dpi=150, bbox_inches="tight")
    print(f"Saved → {OUT_PNG}")


if __name__ == "__main__":
    main()

File: analysis/running_wsls/wsls_scatter_posterior.R
rm(list = ls())

#### SETUP ####
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(patchwork)
library(rstanarm)

DATA_DIR  <- "../../data/ioc-all-fixed-pilot/task"
OUT_PNG   <- "wsls_scatter_posterior.png"

MISUNDERSTOOD <- c(
  "67dae998d8f2cfb8a8e3bf03",
  "6a087d170d5521dd2055e045",
  "6a0ac6bc5fb56e6d7f8e6569"
)
FELT_DIFFERENT <- c(
  "677009b08130c3028f6a8a6d", "68598a1d4cebd213b2abb1d9",
  "69b7e04340b00585acbb91ac", "6a0092581cd317f1ff1765a2",
  "5c41f9ce4fe4f800016dfaac", "651d64e4756ee3358eeb981f",
  "6982544679288685d8a0199f", "69a7048762f2aacbfb3c2f02",
  "69de2b262f16bdb2c0122847", "6a01c4c8248651c0157c76c5"
)
OMITTED <- c(MISUNDERSTOOD, FELT_DIFFERENT)

OI_BLUE   <- "#0072B2"
OI_ORANGE <- "#E69F00"
GREY30    <- "#4d4d4d"
GREY40    <- "#666666"
GREY65    <- "#a6a6a6"
GREY80    <- "#cccccc"

# ── load task data ────────────────────────────────────────────────────────────
files  <- list.files(DATA_DIR, pattern = "^ioc-all_[a-f0-9]{24}_SESSION.*\\.csv$", full.names = TRUE)
df_raw <- lapply(files, \(f) {
  pid <- regmatches(f, regexpr("[a-f0-9]{24}", f))
  read_csv(f, show_col_types = FALSE, col_types = cols(.default = col_character())) |>
    mutate(participant = pid)
}) |> bind_rows()

gb <- df_raw |>
  filter(task == "gambling_choice",
         as.character(block_number) != "training",
         tolower(as.character(is_choice_valid)) == "true") |>
  mutate(
    block_number       = as.integer(block_number),
    trial_number       = as.numeric(trial_number),
    reward             = as.numeric(reward),
    understood_eq_felt = !(participant %in% OMITTED)
  ) |>
  arrange(participant, block_number, trial_number)

# ── per-subject P(stay|win) and P(stay|loss) ──────────────────────────────────
df_scatter <- gb |>
  group_by(participant, understood_eq_felt, block_number) |>
  mutate(
    stay         = as.integer(choice_key == lag(choice_key)),
    reward_nback = lag(reward)
  ) |>
  ungroup() |>
  filter(!is.na(stay), !is.na(reward_nback)) |>
  group_by(participant, understood_eq_felt) |>
  summarise(
    p_win  = mean(stay[reward_nback == 1], na.rm = TRUE),
    p_loss = mean(stay[reward_nback == 0], na.rm = TRUE),
    .groups = "drop"
  ) |>
  filter(!is.na(p_win), !is.na(p_loss))

# ── fit pooled WSLS model for group β posterior ───────────────────────────────
df_pool <- gb |>
  group_by(participant, block_number) |>
  mutate(
    stay         = as.integer(choice_key == lag(choice_key)),
    reward_nback = lag(reward)
  ) |>
  ungroup() |>
  filter(!is.na(stay), !is.na(reward_nback))

fit_pool <- stan_glm(
  stay ~ reward_nback, data = df_pool, family = binomial(),
  prior_intercept = normal(0, 2), prior = normal(0, 2),
  chains = 4, iter = 2000, warmup = 1000, seed = 42, refresh = 0
)
beta_mu <- as.vector(as.matrix(fit_pool)[, "reward_nback"])
med      <- median(beta_mu)
pd       <- max(mean(beta_mu > 0), mean(beta_mu < 0)) * 100
ci90     <- quantile(beta_mu, c(0.05, 0.95))

# ── Panel 1: scatter ──────────────────────────────────────────────────────────
p_scatter <- ggplot(df_scatter, aes(x = p_loss, y = p_win)) +
  geom_abline(slope = 1, intercept = 0, colour = GREY40, linetype = "dashed", linewidth = 0.9, alpha = 0.6) +
  geom_point(aes(colour = understood_eq_felt), size = 2.5, alpha = 0.85) +
  scale_colour_manual(
    values = c("TRUE" = OI_BLUE, "FALSE" = OI_ORANGE),
    labels = c("TRUE" = "understood = felt  (n=22)", "FALSE" = "understood != felt  (n=13)"),
    name   = NULL
  ) +
  scale_x_continuous(limits = c(0, 1)) +
  scale_y_continuous(limits = c(0, 1)) +
  coord_fixed() +
  labs(x = "P(stay | loss)", y = "P(stay | win)") +
  theme_minimal(base_size = 11) +
  theme(
    panel.grid      = element_blank(),
    axis.line.x     = element_line(colour = GREY30),
    axis.line.y     = element_line(colour = GREY30),
    axis.title      = element_text(size = 16),
    axis.text       = element_text(size = 8),
    legend.text     = element_text(size = 10),
    legend.position = "bottom",
    plot.background = element_rect(fill = "white", colour = NA)
  )

# ── Panel 2: β_μ posterior ───────────────────────────────────────────────────
ann_label <- sprintf("median = %.2f\npd = %.1f%%", med, pd)
df_draws  <- data.frame(x = beta_mu)

p_post <- ggplot(df_draws, aes(x = x)) +
  stat_density(aes(y = after_stat(density / max(density))), geom = "area",
               fill = GREY80, colour = NA, bw = "SJ") +
  geom_vline(xintercept = 0,   colour = GREY40, linetype = "dashed", linewidth = 0.9) +
  geom_vline(xintercept = med, colour = GREY65, linetype = "dashed", linewidth = 0.4) +
  annotate("segment", x = ci90[1], xend = ci90[2], y = -0.08, yend = -0.08,
           colour = GREY30, linewidth = 0.9) +
  annotate("point",   x = med, y = -0.08, colour = GREY30, size = 2) +
  annotate("text", x = med, y = 1.22,
           label = ann_label,
           hjust = 0.5, vjust = 1, size = 2.5, colour = "#555555") +
  labs(x = expression(beta[mu] ~ "(reward effect on log-odds of staying)"),
       title = "WSLS reward effect") +
  coord_cartesian(ylim = c(-0.23, 1.5), clip = "off") +
  theme_minimal(base_size = 11) +
  theme(
    panel.grid    = element_blank(),
    axis.title.y  = element_blank(),
    axis.text.y   = element_blank(),
    axis.ticks.y  = element_blank(),
    axis.line.x   = element_line(colour = GREY30),
    axis.title.x  = element_text(size = 16),
    axis.text.x   = element_text(size = 8),
    plot.title    = element_text(size = 16, hjust = 0.5),
    plot.background = element_rect(fill = "white", colour = NA)
  )

# ── combine and save ──────────────────────────────────────────────────────────
fig <- p_scatter | p_post
ggsave(OUT_PNG, fig, width = 10, height = 4.5, dpi = 150, bg = "white")
cat(sprintf("Saved → %s\n", OUT_PNG))
